### PR TITLE
PHPLIB-1122: Support Document and PackedArray objects in public APIs

### PR DESCRIPTION
**Pull request by jmikola**
https://jira.mongodb.org/browse/PHPLIB-1122

Excluding tests, most changes were concentrated in functions.php. A document_to_array() function was introduced to make it easier to access document values as arrays. This is mainly used by other utility functions (e.g. checking for dollar prefixed keys) but also ended up being used in the Find operation for handling the "modifiers" option.

Adds tests for using Document and PackedArray objects in operations.

Beyond handling of document/array types, this also filled gaps in test coverage for:

 * Validation for IndexInput key/name options
 * Passing a MongoDB\Driver\Command directly to DatabaseCommand
 * is_last_pipeline_operator_write()

----

### Addressed

Most changes were made to internal functions, which are used by many operations/classes. The following is a summary of test coverage changes in FunctionsTest:

* `document_to_array()`: various types for `$document` argument. This is used by many other internal functions and the Find operation.
* `is_first_key_operator()`: various types for `$document` argument. This is used by operations that peform updates (e.g. BulkWrite, FindOneAndUpdate, Update).
* `is_last_pipeline_operator_write()`: testing more inputs and using a variety of types for stage documents. This is used by Aggregate (operation and helpers) and indirectly by the Watch operation.
* `is_mapreduce_output_inline()`: various types for `$out` argument. This is used by MapReduce.
* `is_pipeline()`: various types for `$pipeline` argument and stage documents within. This is used for validating update values for `update` and `findAndModify` commands.

The following is a summary of added test coverage for various operations/classes. When an argument/option is listed without additional details, that means we've added test coverage for expressing the document in various forms (i.e. array, stdClass, Serializable, Document):

* CountFunctionalTest:
  * Count: `$filter`
* CountDocumentsFunctionalTest
  * CountDocuments: `$filter`
* CreateIndexesTest:
  * IndexInput: order validation for `key` option. Additionally filled gaps in test coverage for validating `key` and `name` options (unrelated to documents)
* CreateIndexesFunctionalTest:
  * CreateIndexes: `key` option
  * IndexInput: generating `name` option from `key` option
* DatabaseCommand
  * DatabaseCommand: `$command`. Additionally added test coverage for passing a `MongoDB\Driver\Command` object constructed from various document types.
* DeleteFunctionalTest
  * DeleteMany: `$filter`
  * DeleteOne: `$filter`
* DistinctFunctionalTest
  * Distinct: `$filter`
* FindFunctionalTest
  * Find: `$filter`, `modifiers` option
  * FindOne: `$filter`, `modifiers` option
* FindAndModifyFunctionalTest
  * FindAndModify: `filter` and `query` options
  * FindOneAndDelete: `$filter`
  * FindOneAndReplace: `$filter`, `$replacement`
  * FindOneAndUpdate: `$filter`, `$update`
* FindOneAndReplaceTest
  * FindOneAndReplace: `$replacement` validation
* FindOneAndUpdateTest
  * FindOneAndUpdate: `$update` validation
* InsertManyFunctionalTest
  * InsertMany: encoding `$documents` elements and `_id` generation/access
* InsertOneFunctionalTest
  * InsertOne: encoding `$document` and `_id` generation/access
* ReplaceOneTest
  * ReplaceOneTest: `$replacement` validation
* UpdateTest
  * Update: `$update` validation based on `multi` option
* UpdateFunctionalTest
  * Update: `$filter`, `$update`
  * ReplaceOne: `$filter`, `$replacement`
  * UpdateMany: `$filter`, `$update`
  * UpdateOne: `$filter`, `$update`
* UpdateManyTest
  * UpdateMany: `$update` validation
* UpdateOneTest
  * UpdateOne: `$update` validation

### Omitted

The following are notable parts of PHPLIB that were _not_ touched by this PR:

* `apply_type_map_to_document()`: only used in GridFS on stdClass objects
* GridFS metadata document: not used beyond type validation as an "array or object"

### CODE REVIEW

**Review by alcaeus — COMMENTED** [review]
That's a lot of tests being added, and it looks like you've covered all of the new use cases for packed arrays and documents. I left a few suggestions that should also help with getting psalm to play nice with these changes.

File: src/functions.php [hunk 1]
```diff
@@ -208,6 +235,21 @@ function is_first_key_operator($document): bool
...
+         * iterator_to_array() since Document iteration returns all values as
+         * MongoDB\BSON\Value instances. */
+        $pipeline = $pipeline->toPHP([
+            'array' => 'bson',
+            'document' => 'bson',
+            'root' => 'array',
+        ]);
+    }
+
+    if ($pipeline instanceof Serializable) {
```
[the remark this answers is not in the document]
Comment (GromNaN) [reply src/functions.php:249]: I would rather use `elseif`.

File: src/functions.php [hunk 2]
```diff
@@ -272,9 +314,15 @@ function is_last_pipeline_operator_write(array $pipeline): bool
...
+    if (! is_array($lastOp) && ! is_object($lastOp)) {
+        return false;
+    }
+
+    $lastOp = document_to_array($lastOp);
+
+    reset($lastOp);
 
-    return in_array(key($lastOp), ['$out', '$merge'], true);
+    return key($lastOp) === '$merge' || key($lastOp) === '$out';
```
[the remark this answers is not in the document]
Comment (GromNaN) [reply src/functions.php:325]: We should use rector to track that automatically. https://github.com/rectorphp/rector/blob/main/docs/rector_rules_overview.md#php73

File: src/functions.php [hunk 3]
```diff
@@ -94,15 +95,34 @@ function apply_type_map_to_document($document, array $typeMap)
...
+ * type checking. Prohibiting PackedArray would only require callers to check
+ * for a PackedArray in order to avoid an exception. Furthermore, this function
+ * does not distinguish between Serializable::bsonSerialize() return values that
+ * might encode as a BSON document or array.
  *
  * @internal
- * @param array|object $document Document containing fields mapped to values,
- *                               which denote order or an index type
- * @throws InvalidArgumentException
+ * @param array|object $document
```
[the remark this answers is not in the document]
Comment (GromNaN) [reply src/functions.php:110]: The phpdoc is about what is expected. It's not a validation function that should accept anything and fail in case of invalid input. We need to throw a `TypeError`. When upgrading to PHP 8.0+ the union type will move to the code.

File: src/functions.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (jmikola) [reply src/functions.php:249]: This thread just made me wonder what would happen if [`Serializable::bsonSerialize()`](https://www.php.net/manual/en/mongodb-bson-serializable.bsonserialize.php) were to return a PackedArray or Document, but I remembered that's [not possible](https://github.com/mongodb/mongo-php-driver/blob/1.15.3/src/phongo_bson_encode.c#L114).

Will refactor this and similar code in `document_to_array()` to use `elseif`.

File: src/functions.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (jmikola) [reply src/functions.php:325]: > Consider using [array_key_first](https://www.php.net/manual/en/function.array-key-first.php) (available as of PHP 7.3.0, symfony/polyfill might be required) instead of resetting the array pointer. This likely applies to a number of other usages as well.

TIL those functions exist. Will do!

> We should use rector to track that automatically. https://github.com/rectorphp/rector/blob/main/docs/rector_rules_overview.md#php73

@GromNaN: Would you mind creating a PHPLIB ticket to propose adding Rector to our CI?

File: src/functions.php [hunk 4]
```diff
@@ -285,25 +333,14 @@ function is_last_pipeline_operator_write(array $pipeline): bool
  * @internal
  * @see https://mongodb.com/docs/manual/reference/command/mapReduce/#output-inline
  * @param string|array|object $out Output specification
- * @throws InvalidArgumentException
```
[the remark this answers is not in the document]
Comment (jmikola) [reply src/functions.php:288]: I originally removed this because `is_mapreduce_output_inline()` starts with:

```php
if (! is_array($out) && ! is_object($out)) {
    return false;
}
```

`document_to_array()` only throws if the final `$document` value is not an array of object, so I assumed an exception wouldn't be possible.

Thinking about this some more, I suppose we could still throw if a Serializable instance were to return something other than an array or object from `bsonSerialize()`. Granted, that would break in other ways, but that's beyond the scope of ``is_mapreduce_output_inline()`.

In that case, it's still possible for `is_pipeline()` to throw for the same reason (i.e. a Serializable object in the pipeline array returns a bad value).

Note: I think this edge case only applies to PHP < 7.x, since `bsonSerialize()` has a return type hint for PHP 8.0: https://github.com/mongodb/mongo-php-driver/blob/1.15.3/src/BSON/Serializable.stub.php#L12-L18

File: src/functions.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (jmikola) [reply src/functions.php:110]: > The phpdoc is about what is expected.

Correct. Only arrays or objects should be provided to `document_to_array()`.

> We need to throw a TypeError.

PHPLIB's own type checking uses `InvalidArgumentException::invalidType()` everywhere. I suggest keeping that here for consistency. When we can rely on PHP itself to enforce the parameter type, then we'll get TypeError.

That said, both of these exceptions should be very unlikely for the following reasons:

 * Excluding test cases, every code path that invokes `document_to_array()` should already have validated the `$document` before doing so.
 * The edge case for `document_to_array()` being passed an object and still throwing has to do with `Serializable::bsonSerialize()` returning a non-array, non-object. That should only be possible on PHP 7.x, since [8.0+ declares a union return type hint](https://github.com/mongodb/mongo-php-driver/blob/1.15.3/src/BSON/Serializable.stub.php#L12-L18).

File: tests/Operation/CountDocumentsFunctionalTest.php [hunk 5]
```diff
@@ -2,11 +2,46 @@
...
+    public function provideFilterDocuments(): array
+    {
+        $expectedQuery = (object) ['x' => 1];
+
+        return [
+            'array' => [['x' => 1], $expectedQuery],
+            'object' => [(object) ['x' => 1], $expectedQuery],
+            'Serializable' => [new BSONDocument(['x' => 1]), $expectedQuery],
+            'Document' => [Document::fromPHP(['x' => 1]), $expectedQuery],
+        ];
```
[the remark this answers is not in the document]
Comment (jmikola) [reply tests/Operation/CountDocumentsFunctionalTest.php:42]: Created [PHPLIB-1139](https://jira.mongodb.org/browse/PHPLIB-1139) as follow-up.

File: tests/Operation/CreateIndexesFunctionalTest.php [hunk 6]
```diff
@@ -119,6 +122,20 @@ public function testCreateIndexes(): void
...
+    public function provideKeyCasts(): array
+    {
+        // phpcs:disable SlevomatCodingStandard.ControlStructures.JumpStatementsSpacing
+        // phpcs:disable Squiz.Functions.MultiLineFunctionDeclaration
+        // phpcs:disable Squiz.WhiteSpace.ScopeClosingBrace.ContentBefore
+        return [
+            'array' => [function ($key) { return $key; }],
+            'object' => [function ($key) { return (object) $key; }],
+            'Serializable' => [function ($key) { return new BSONDocument($key); }],
+            'Document' => [function ($key) { return Document::fromPHP($key); }],
```
[the remark this answers is not in the document]
Comment (jmikola) [reply tests/Operation/CreateIndexesFunctionalTest.php:134]: Indeed. I lamented not being able to make use of [arrow functions](https://www.php.net/manual/en/functions.arrow.php) while writing these.

File: src/functions.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (jmikola) [reply src/functions.php:288]: The comment above is not relevant.

See: https://github.com/mongodb/mongo-php-library/pull/1077#discussion_r1205074612

File: src/functions.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (alcaeus) [reply src/functions.php:110]: Ah, I see. I saw this as a validation function which would take anything. I can't wait until we can bump the PHP requirement to PHP 8.2 (or at least 8.1 to get intersection types) and make full use of the PHP type system.

File: src/functions.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (GromNaN) [reply src/functions.php:325]: Of course: https://jira.mongodb.org/browse/PHPLIB-1141

**Review by GromNaN — APPROVED** [review]

**Review by GromNaN — APPROVED** [review]
Take my approval for what it is worth. SGTM.

File: src/functions.php [hunk 7]
```diff
@@ -94,15 +95,34 @@ function apply_type_map_to_document($document, array $typeMap)
...
+ * @throws InvalidArgumentException if $document is not an array or object
  */
-function generate_index_name($document): string
+function document_to_array($document): array
 {
+    if ($document instanceof Document || $document instanceof PackedArray) {
+        /* Nested documents and arrays are intentionally left as BSON. We avoid
+         * iterator_to_array() since Document and PackedArray iteration returns
+         * all values as MongoDB\BSON\Value instances. */
+        return $document->toPHP([
```
[the remark this answers is not in the document]
Comment (jmikola) [reply src/functions.php:119]: Done. I found a few more errors as well and fixed all but one (see the commit for the regenerated baseline).

**Review by alcaeus — APPROVED** [review]
LGTM. I've rebased this PR on master, and as long as the psalm check is happy, I'm happy. While the baseline contains lots of changes, I believe most of those are from a previous generation on PHP 8+ (I believe I last regenerated the baseline on PHP 8.2).

**Comment by jmikola** [comment]
CI failures are related to CSFLE QEv2 protocol changes, and two timeouts in CSFLE tests on Debian 9.2 (unrelated to this PR). I'll proceed with merging.
